drama-queen/src/core/tools/externalResources.ts

```diff
@@ -0,0 +1,129 @@
+  const questionnairesWrapper = await fetchUrl<ExternalQuestionnairesWrapper>({
+    url: `${externalResourcesUrl}/gide-questionnaires.json`,
+  })
+
+  return questionnairesWrapper.questionnaires
+}
+
+// Get the list of external resource URLs for a questionnaireId
+export async function getTransformedManifest(
+  questionnaireId: string
+): Promise<string[]> {
+  // get the manifest for a questionnaireId
+  const manifest = await fetchUrl<Manifest>({
+    url: `${externalResourcesUrl}/${questionnaireId}/assets-manifest.json`,
+  })
+
+  // Transform the manifest values into resource URLs, and get an array of these resource URLs
+  const transformedManifest = Object.values(manifest).map(
+    (resourceUrl) => `${externalResourcesUrl}/${resourceUrl}`
+  )
+
+  return transformedManifest
+}
+
+// Filter resources from manifest that are already cached, to avoid useless requests (overly large files).
+export async function filterTransformedManifest(
+  cacheName: string,
+  transformedManifest: string[]
+): Promise<string[]> {
+  // Open the specified cache
+  const cacheForManifest = await caches.open(cacheName)
+
+  // If cache is not available, return a copy of transformedManifest
+  if (!cacheForManifest) {
+    return [...transformedManifest]
+  }
+
+  // Get all urls from the cache
+  const cachedUrls = (await cacheForManifest.keys()).map(
+    (request) => request.url
+  )
+
+  // filter the transformedManifest for keeping only files that are not cached yet
+  const uncachedResources = transformedManifest.filter(
+    (resourceUrl) => !cachedUrls.includes(resourceUrl)
+  )
+
+  return uncachedResources
+}
+
+// Cache every external resources (not already cached) for a particular questionnaire
+export async function getResourcesFromExternalQuestionnaire(
+  questionnaire: ExternalQuestionnaire
+): Promise<void> {
+  const transformedManifest = await getTransformedManifest(questionnaire.id)
+
+  const filteredTransformedManifest = await filterTransformedManifest(
+    questionnaire.cacheName,
+    transformedManifest
+  )
+
+  const manifestCache = await caches.open(questionnaire.cacheName)
+  return await manifestCache.addAll(filteredTransformedManifest)
+}
+
+// Separate, from the list of external questionnaires, those that are needed and those that are not needed
+export function getExternalQuestionnaireFiltered(
+  neededQuestionnaireIds: string[],
+  externalQuestionnaires: ExternalQuestionnaires
+): ExternalQuestionnairesFiltered {
+  return externalQuestionnaires.reduce(
+    (result, questionnaire) => {
+      // Check if the current questionnaire's id matches any of the needed IDs
+      const isNeeded = neededQuestionnaireIds.some((neededId) =>
+        neededId.toLowerCase().includes(questionnaire.id.toLowerCase())
+      )
+
+      // If the external questionnaire is needed
+      if (isNeeded) {
+        result.neededQuestionnaires.push(questionnaire)
```

Comment:
You should make a copy of result and not alter it directly. I think it's bad practice and can create bug here.

Reply:
Ok you should not make a copy as it would decrease performance and reduce is (one of) the only case where immutability is thrown off the window BUT instead maybe you should use a for loop instead of reduce to increase performance and readability. (see mdn documentation on reduce method and various benchmarks)